File: kotlin/src/com/adviser/ipaddress/TestIpv6Mapped.kt
package com.adviser.ipaddress

import java.math.BigInteger
import kotlin.test.assertEquals

class IPv6MappedTest(IPAddress ip, String s, String sstr,
    String string, BigInteger u128, String address) {
    public IPAddress ip
    public String s
    public String sstr
    public String string
    public BigInteger u128
    public String address
    HashMap<String, BigInteger> valid_mapped = HashMap<String, BigInteger>()
    HashMap<String, BigInteger> valid_mapped_ipv6 = HashMap<String, BigInteger>()
    HashMap<String, String> valid_mapped_ipv6_conversion = HashMap<String, String>()
        this.ip = ip
        this.s = s
        this.sstr = sstr
        this.string = string
        this.u128 = u128
        this.address = address
    }
}

class TestIpv6Mapped {

    public fun IPv6MappedTest setup() {
        val ret = IPv6MappedTest(
            Ipv6Mapped.create("::172.16.10.1").unwrap(),
            "::ffff:172.16.10.1",
            "::ffff:172.16.10.1/32",
            "0000:0000:0000:0000:0000:ffff:ac10:0a01/128",
            BigInteger("281473568475649"),
            "::ffff:ac10:a01/128");
        ret.valid_mapped.put("::13.1.68.3", BigInteger("281470899930115"));
        ret.valid_mapped.put("0:0:0:0:0:ffff:129.144.52.38",
                            BigInteger("281472855454758"));
        ret.valid_mapped.put("::ffff:129.144.52.38",
                            BigInteger("281472855454758"));
        ret.valid_mapped_ipv6.put("::ffff:13.1.68.3", BigInteger("281470899930115"));
        ret.valid_mapped_ipv6.put("0:0:0:0:0:ffff:8190:3426",
                                 BigInteger("281472855454758"));
        ret.valid_mapped_ipv6.put("::ffff:8190:3426",
                                 BigInteger("281472855454758"));
        ret.valid_mapped_ipv6_conversion.put("::ffff:13.1.68.3", "13.1.68.3");
        ret.valid_mapped_ipv6_conversion.put("0:0:0:0:0:ffff:8190:3426", "129.144.52.38");
        ret.valid_mapped_ipv6_conversion.put("::ffff:8190:3426", "129.144.52.38");
        return ret
    }


    @Test
    public fun test_initialize() {
        val s = setup();
        assertEquals(true, IPAddress.parse("::172.16.10.1").isOk());
        s.valid_mapped.forEach[ip, u128 |
            //println!("-{}--{}", ip, u128);
            //if IPAddress.parse(ip).is_err() {
            //    println!("{}", IPAddress.parse(ip).unwrapErr());
            //}
            assertEquals(true, IPAddress.parse(ip).isOk());
            assertEquals(u128, IPAddress.parse(ip).unwrap().host_address);
        ]
        s.valid_mapped_ipv6.forEach[ip, u128 |
            //println!("===={}=={:x}", ip, u128);
            assertEquals(true, IPAddress.parse(ip).isOk());
            assertEquals(u128, IPAddress.parse(ip).unwrap().host_address);
        ]
    }
    @Test
    public fun test_mapped_from_ipv6_conversion() {
        setup().valid_mapped_ipv6_conversion.forEach[ip6, ip4 |
            //println!("+{}--{}", ip6, ip4);
            assertEquals(ip4, IPAddress.parse(ip6).unwrap().mapped.to_s());
        ]
    }
    @Test
    public fun test_attributes() {
        val s = setup();
        assertEquals(s.address, s.ip.to_string());
        assertEquals(128, s.ip.prefix.num);
        assertEquals(s.s, s.ip.to_s_mapped());
        assertEquals(s.sstr, s.ip.to_string_mapped());
        assertEquals(s.string, s.ip.to_string_uncompressed());
        assertEquals(s.u128, s.ip.host_address);
    }
    @Test
    public fun test_method_ipv6() {
        assertTrue(setup().ip.is_ipv6());
    }
    @Test
    public fun test_mapped() {
        assertTrue(setup().ip.is_mapped());
    }
}

File: js/test/test_ipv6_mapped.ts
import IPAddress from "../src/ipaddress.js";
// import Prefix128 from '../src/prefix128';
// import Ipv6Unspec from '../src/ipv6_unspec';
import Ipv6Mapped from "../src/ipv6_mapped.js";
import Crunchy from "../src/crunchy.js";

interface IPv6MappedTestParams {
  ip: IPAddress;
  s: string;
  sstr: string;
  string: string;
  u128: Crunchy;
  address: string;
}

class IPv6MappedTest {
  ip: IPAddress;
  s: string;
  sstr: string;
  string: string;
  u128: Crunchy;
  address: string;
  valid_mapped: [string, Crunchy][] = [];
  valid_mapped_ipv6: [string, Crunchy][] = [];
  valid_mapped_ipv6_conversion: [string, string][] = [];
  constructor(obj: IPv6MappedTestParams) {
    this.ip = obj.ip;
    this.s = obj.s;
    this.sstr = obj.sstr;
    this.string = obj.string;
    this.u128 = obj.u128;
    this.address = obj.address;
  }
}

describe("ipv6_mapped", () => {
  function setup(): IPv6MappedTest {
    const ipv6 = new IPv6MappedTest({
      ip: Ipv6Mapped.create("::172.16.10.1"),
      s: "::ffff:172.16.10.1",
      sstr: "::ffff:172.16.10.1/32",
      string: "0000:0000:0000:0000:0000:ffff:ac10:0a01/128",
      u128: Crunchy.parse("281473568475649"),
      address: "::ffff:ac10:a01/128",
    });
    ipv6.valid_mapped.push(["::13.1.68.3", Crunchy.parse("281470899930115")]);
    ipv6.valid_mapped.push([
      "0:0:0:0:0:ffff:129.144.52.38",
      Crunchy.parse("281472855454758"),
    ]);
    ipv6.valid_mapped.push([
      "::ffff:129.144.52.38",
      Crunchy.parse("281472855454758"),
    ]);
    ipv6.valid_mapped_ipv6.push([
      "::ffff:13.1.68.3",
      Crunchy.parse("281470899930115"),
    ]);
    ipv6.valid_mapped_ipv6.push([
      "0:0:0:0:0:ffff:8190:3426",
      Crunchy.parse("281472855454758"),
    ]);
    ipv6.valid_mapped_ipv6.push([
      "::ffff:8190:3426",
      Crunchy.parse("281472855454758"),
    ]);
    ipv6.valid_mapped_ipv6_conversion.push(["::ffff:13.1.68.3", "13.1.68.3"]);
    ipv6.valid_mapped_ipv6_conversion.push([
      "0:0:0:0:0:ffff:8190:3426",
      "129.144.52.38",
    ]);
    ipv6.valid_mapped_ipv6_conversion.push([
      "::ffff:8190:3426",
      "129.144.52.38",
    ]);
    return ipv6;
  }

  it("test_initialize", () => {
    const s = setup();
    assert.isOk(IPAddress.parse("::172.16.10.1"));
    for (const i of s.valid_mapped) {
      const ip = i[0];
      const u128 = i[1];
      // println("-{}--{}", ip, u128);
      if (!IPAddress.parse(ip)) {
        // console.log(IPAddress.parse(ip));
      }
      assert.isOk(IPAddress.parse(ip));
      assert.equal(
        u128.toString(),
        IPAddress.parse(ip).host_address.toString(),
      );
    }
    for (const i of s.valid_mapped_ipv6) {
      const ip = i[0];
      const u128 = i[1];
      // println("===={}=={:x}", ip, u128);
      assert.isOk(IPAddress.parse(ip));
      assert.equal(
        u128.toString(),
        IPAddress.parse(ip).host_address.toString(),
      );
    }
  });
  it("test_mapped_from_ipv6_conversion", () => {
    for (const i of setup().valid_mapped_ipv6_conversion) {
      const ip6 = i[0];
      const ip4 = i[1];
      assert.equal(ip4, IPAddress.parse(ip6).mapped.to_s());
    }
  });
  it("test_attributes", () => {
    const s = setup();
    assert.equal(s.address, s.ip.to_string());
    assert.equal(128, s.ip.prefix.num);
    assert.equal(s.s, s.ip.to_s_mapped());
    assert.equal(s.sstr, s.ip.to_string_mapped());
    assert.equal(s.string, s.ip.to_string_uncompressed());
    assert.equal(s.u128.toString(), s.ip.host_address.toString());
  });
  it("test_method_ipv6", () => {
    assert.isOk(setup().ip.is_ipv6());
  });
  it("test_mapped", () => {
    assert.isOk(setup().ip.is_mapped());
  });
});
